### ISSUE TITLE
[Feature] Unify the MoE router GEMM behind one gate layer

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#38695]
## Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Motivation

Router GEMM precision is critical, and it matters most on long multi-turn work, where small differences accumulate over a long CoT. For example, in #29892 the routing precision change was worth +12.6pp resolved on swe-bench Verified (66.8% to 79.4%, 500 problems, GLM-5.2-FP8 on 4xGB300), even though the per-layer logit difference was only 2.6e-7 and single-layer top-k was bit-identical over 400 samples.

The precision that matters is in the output store. cuBLAS already accumulates a bf16 GEMM in fp32, so asking for fp32 output keeps the accumulator and costs nothing. For example, in #29783 we found that doing this on DSv3 moved accept length from 3.413 to 3.894.

Upcasting the activation as well is a much more expensive way to get there. In #29892 that path took 3 kernels at 16.7us per layer against 1 kernel at 8.7us, which over 93 MoE layers came out to 292 tok/s versus 327 tok/s, so -10.5% decode. That PR was closed once we agreed the bf16 GEMM with fp32 output gives the same effect.

Currently, every model in sglang implements its own version of this. For example, `deepseek_v2` has a six-way branch (AMX, deterministic, prefill-CP, tiny-gemm, aiter, cuBLAS), while `nemotron_h` and `minimax_m3` each open-code a bf16 matmul with fp32 output. Meanwhile GLM-4 MoE, `mimo_v2`, `laguna`, `hunyuan_v3`, `sarvam_moe` and `step3p5` still run the expensive activation upcast, so they are paying roughly what #29892 measured for an effect we can get for free. Similarly, `bailing_moe_v3` uses a triton matvec below M=8 and `deepseek_v2` uses tiny-gemm below M=16, so we have two thresholds and two kernels with slightly different numerics, and neither is shared.

Because of this, fixes do not propagate. `mimo_v2` was branched from `deepseek_v2` before #29783 landed, so it still casts the correction bias to bf16. `ernie4` and `kimi_linear` have the same problem, except unconditionally. And `bailing_moe`, `bailing_moe_linear` and `llada2` compute the router in fp32 and then cast the result straight back to bf16, which throws away the precision they just paid for. All of these wrote their own gate. K3 and K2.5 reuse deepseek's, so they picked up the fix for free.

We already have the kernels for this. #34693 already folded the DSv3 router kernel into the shared tiny GEMM for `deepseek_v2`, `dots3` and `kimi_k3`, so the kernel side is consolidating. The bf16 to fp32 GEMM helper in `kernels/ops/attention/dsv4/gemm.py` is also already a four-tier dispatcher (aiter, HPC bf16 split, DeepGEMM, cuBLAS). However, it lives under an attention directory named after one model, and longcat is the only model that opts into the HPC tier or tunes its threshold. Everyone else silently falls through to cuBLAS.

## Proposal

We should add a `RouterGate` layer next to `srt/layers/moe/topk.py` that always hands top-k fp32 logits, computed in bf16. The layer owns the kernel and the output dtype. This is also roughly what vLLM does, where the checkpoint config decides precision and the gate layer picks the kernel.

Deterministic inference should also be handled in the layer. Currently only `deepseek_v2` honors it for the router, so for example `bailing_moe_v3` still switches kernel at small token counts with no guard, and the same token can route differently depending on batch size even with the flag on.

As part of this we should move the bf16 to fp32 GEMM helper out of the DSv4 attention directory, since the router is its main consumer, and move the small-token-count threshold helper out of `models/deepseek_common/`. It currently lives in a model package, which is probably why bailing wrote its own.

Independent of the refactor, there are also three small fixes we can land now: drop the bf16 correction bias in `mimo_v2`, `ernie4` and `kimi_linear`, and drop the cast back to bf16 in `bailing_moe`, `bailing_moe_linear` and `llada2`.

## Benefits

- Every model gets fp32 router logits at the cost of the bf16 GEMM.
- A router numerics fix lands in one place and applies everywhere.
- The HPC and DeepGEMM tiers become available to every model.
- The small-token-count path has one threshold and one kernel.

## Related resources

- #29783, the original DSv3 router fix
- #29892, where we measured the activation upcast and closed it in favor of this approach
- #31001, #31502 and #31927, where the bf16 bias cast was re-added and then removed again after the FlashInfer 0.6.15 bump
- #34693, which replaced the DSv3 router GEMM with the unified tiny GEMM
- Megatron-LM's routing precision change for fine-grained MoE: https://github.com/NVIDIA/Megatron-LM/commit/dab7723821fc326564634b398a809d43740a6c8d
- A recent silent-downcast fix in their ModelOpt example: https://github.com/NVIDIA/Megatron-LM/pull/5810
- vLLM's `GateLinear`: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/router/gate_linear.py

**Comment by b8zhong** [comment]
We can do something like this: unconditional fp32 output, and whether we need fp32 inputs too. For validated models we can use `fp32_compute=False` (like `deepseek_v2`/`nemotron_h`/etc)

```python
self.gate = RouterGate(
    hidden_size=config.hidden_size,
    num_experts=config.n_routed_experts,
    has_correction_bias=True,
    fp32_compute=True,  # default
)
```

**Comment by aayush7511** [comment]
Hi, can I work on this

**Comment by DengZhiyuan-math** [comment]
I’d like to take the independent small fix for `bailing_moe`, `bailing_moe_linear`, and `llada2`, not the `RouterGate` refactor. I propose preserving FP32 router logits by removing the final cast back to `hidden_states.dtype` before top-k. I’ll add a focused synthetic regression, parametrized across the three gates, that verifies FP32 output and retains separation for close scores that collapse in BF16.

**Comment by b8zhong** [comment]
Hi, @DengZhiyuan-math Sure! There's no need to add unit tests. It's only fixing the modelling file.

**Comment by DengZhiyuan-math** [comment]
Thanks @b8zhong for confirming, and for clarifying that unit tests are not needed for this model-file fix.

I noticed that @seongyun1104's open PR #38719 already covers the router-logit downcast removal in `bailing_moe`, `bailing_moe_linear`, and `llada2`, which overlaps with the small fix I proposed above.

Would you prefer that I coordinate with the author to split out a minimal fix for these three models, or help with the existing PR? I'd like to avoid duplicating their work. My proposed scope remains the three-model fix, separate from the broader `RouterGate` refactor.
